bluesky_widgets/models/run_engine_client.py

```diff
@@ -0,0 +1,372 @@
+import time
+
+from bluesky_live.list import ListModel
+from bluesky_live.event import EmitterGroup, Event
+from bluesky_queueserver.manager.comms import ZMQCommSendThreads, CommTimeoutError
+
+
+class PlanItem:
+    def __init__(self, name, args):
+        self._name = name
+        self._args = args
+        self.events = EmitterGroup(
+            source=self,
+            name=Event,
+        )
+
+    @property
+    def name(self):
+        return self._name
+
+    @name.setter
+    def name(self, value):
+        if value == self._name:
+            return
+        self._name = value
+        self.events.name(name=value)
+
+    @property
+    def args(self):
+        return self._args
+
+    @args.setter
+    def args(self, value):
+        # TODO Deal with *mutation* (editing) of the args the same way we deal
+        # with mutation of plot styles.
+        if value == self._args:
+            return
+        self._args = value
+        self.events.args(args=value)
+
+
+class PlanQueue(ListModel):
+    pass
+
+
+class PlanHistory(ListModel):
+    pass
+
+
+class BigModel:
+    def __init__(self):
+        self._client = RunEngineClient()
+        self.plan_queue = PlanQueue()
+        self.plan_queue.events.adding.connect(self._on_plan_added)
+        self.plan_history = PlanHistory()
+
+    def _on_plan_added(self, event):
+        plan_item = event.item
+        self._client.add(plan_item.name, plan_item.args)
+
+
+class RunEngineClient:
+    def __init__(self, worker_address=None):
+        self._client = ZMQCommSendThreads(zmq_server_address=worker_address)
+
+        self._re_manager_status = {}
+        self._re_manager_accessible = None
+        self._re_manager_status_time = time.time()
+        # Minimum period of status update (avoid excessive call frequency)
+        self._re_manager_status_update_period = 0.2
+
+        self.events = EmitterGroup(
+            source=self,
+            status_changed=Event,
+        )
+
+    @property
+    def re_manager_status(self):
+        return self._re_manager_status
+
+    @property
+    def re_manager_accessible(self):
+        return self._re_manager_accessible
+
+    def clear(self):
+        # Clear the queue.
+        response = self._client.send_message(method="queue_clear")
+        if not response["success"]:
+            raise RuntimeError(f"Failed to clear the plan queue: {response['msg']}")
+
+    def clear_online_status(self):
+        self._re_manager_accessible = None
+        self.events.status_changed()
```

Comment:
It might be convenient for the callbacks to receive the `status` in the `Event`. (Any keyword arguments are allowed here, and then become attributes on the emitted `Event`.)

```suggestion
        self._re_manager_status.clear()
        self.events.status_changed(status=self._re_manager_status)
```